### ISSUE TITLE
The Collapse of the Unix Philosophy

### ISSUE THREAD

**Issue by lucaswadedavis** [lucaswadedavis/reading-list#42]
The Collapse of the Unix Philosophy<br><br>
http://ift.tt/2lZ8F8c<br><br><div>
<div class="post-body">
<p>In the first part of the article, I will enumerate lots of UNIX cheap and dirty hacks, and other various drawbacks. In the second part, we&rsquo;ll talk about the UNIX philosophy. This article was written hastily, and I don&rsquo;t want to further improve it. You&rsquo;re lucky I wrote it. Therefore, I may provide some facts without source links.</p>
<p>Dirty hacks in UNIX started to arise when UNIX was released, and it was long before Windows came to the scene, I guess there wasn&rsquo;t even Microsoft DOS at the time (I guess and I don&rsquo;t bother to check, so check it yourself). If you don&rsquo;t feel like reading, at least look through all the points. You&rsquo;ll find something interesting. It&rsquo;s not a complete list. These are simply the flaws I wanted to mention.</p>
<ul>
<li>
<p>At the very beginning, <code>make</code> was a program one person wrote for himself and some of his friends. Without thinking twice, he made commands understand strings that start with a tab. I.e., a tab was treated differently from a space, which was extremely bad and unusual for UNIX and everywhere beyond it. He did it this way, as he didn&rsquo;t think that <code>make</code> would be used by anyone except for this small group of people. Later came the idea that <code>make</code> was a good thing and it would be nice to include it into the standard UNIX package. <em>In order not to break the already written makefiles, (meaning written by these ten people), he decided not to change anything</em>. Well, that&rsquo;s how it goes&hellip; We all suffer because of those ten guys.</p>
</li>
<li>
<p>Almost at the very beginning, there was no <em>/usr</em> folder in UNIX. All binaries were located in <em>/bin</em> and <em>/sbin</em>. Later, all the information could no longer fit on the disc that was at the disposal of UNIX authors (Dennis Ritchie and Ken Thompson). Therefore, they took another disc and created <em>/usr</em> folder, and another <em>bin</em>, and another <em>sbin</em> in it. They mounted the new disc to <em>/usr</em>. That&rsquo;s where it went from. That&rsquo;s how &ldquo;the second hierarchy&rdquo; of <em>/usr</em> appeared. At some moment, there also appeared &ldquo;the third hierarchy&rdquo; of <em>/usr/local</em>, and then <em>/opt</em>. The narrator of the story said &ldquo;I&rsquo;m still waiting for <em>/opt/local</em> to show up&hellip;&rdquo; <a href="http://ift.tt/y2qeVP">Here&rsquo;s where</a> I learned the story. It&rsquo;s a more accurate version of what happened.</p>
</li>
<li>
<p><em>sbin</em> originally meant &ldquo;static bin&rdquo;, rather than &ldquo;superuser bin&rdquo; as one might think. <em>sbin</em> used to contain static binaries. But then <em>sbin</em> began to contain dynamic binaries, and its name lost its meaning.</p>
</li>
<li>
<p>Windows is often criticized for having a registry. At the same time, people say the approach of UNIX-like systems (tons of configs) is better. By the way, once there appeared a &ldquo;feature&rdquo; in ext4, (it&rsquo;s a big question whether it&rsquo;s a bug), that made Gnome loose all its configs in the user&rsquo;s working directory after a power loss. In the discussion of the bug report, ext4&rsquo;s maintainer said that Gnome&rsquo; should have used something like a registry for storing the information. The sources: <a href="http://ift.tt/2lymVRG">one</a> and <a href="http://ift.tt/2ljdWbR">two</a>. The name of ext4&rsquo;s maintainer is Theodore Ts&rsquo;o. Here&rsquo;s what he said:</p>
</li>
</ul>
<blockquote>
<p>If you really care about making sure something is on disk, you have to use fsync or fdatasync. If you are about the performance overhead of <code>fsync()</code>, <code>fdatasync()</code> is much less heavyweight, if you can arrange to make sure that the size of the file doesn&rsquo;t change often. You can do that via a binary database, that is grown in chunks, and rarely truncated. I&rsquo;ll note that I use the GNOME desktop (which means the gnome panel, but I&rsquo;m not a very major desktop user), and <code>find .[a-zA-Z]\* -mtime 0</code> doesn&rsquo;t show a large number of files. I&rsquo;m guessing it&rsquo;s certain badly written applications which are creating the &laquo;hundreds of dot files&raquo; that people are reporting become zero lengh, and if they are seeing it happen a lot, it must be because the dot files are getting updated very frequently. I don&rsquo;t know what the bad applications are, but the people who complained about large number of state files disappearing should check into which application were involved, and try to figure out how often they are getting modified. As I said, if large number of files are getting frequently modified, it&rsquo;s going to be bad for SSD&rsquo;s as well, there are multiple reasons to fix badly written applications, even if 2.6.30 will have a fix for the most common cases. (Although some server folks may mount with a flag to disable it, since it will cost performance.)</p>
</blockquote>
<p>That&rsquo;s not to mention the fact that critical UNIX files (such as <em>/etc/passwd</em>) that are read upon every (!) call, say, <code>ls -l</code>, are plain text files. The system reads and parses these files again and again, after every single call! It would be much better to use a binary format. Or a database. Or a registry. At least for system files that are critical for performance.</p>
<blockquote>
<p>Two famous people, one from MIT and another from Berkeley (but working on Unix) once met to discuss operating system issues. The person from MIT was knowledgeable about ITS (the MIT AI Lab operating system) and had been reading the Unix sources. He was interested in how Unix solved the PC loser-ing problem. The PC loser-ing problem occurs when a user program invokes a system routine to perform a lengthy operation that might have significant state, such as IO buffers. If an interrupt occurs during the operation, the state of the user program must be saved. Because the invocation of the system routine is usually a single instruction, the PC of the user program does not adequately capture the state of the process. The system routine must either back out or press forward. The right thing is to back out and restore the user program PC to the instruction that invoked the system routine so that resumption of the user program after the interrupt, for example, re-enters the system routine. It is called &ldquo;PC loser-ing&rdquo; because the PC is being coerced into &ldquo;loser mode&rdquo;, where &ldquo;loser&rdquo; is the affectionate name for &ldquo;user&rdquo; at MIT. The MIT guy did not see any code that handled this case and asked the New Jersey guy how the problem was handled. The New Jersey guy said that the Unix folks were aware of the problem, but the solution was for the system routine to always finish, but sometimes an error code would be returned that signaled that the system routine had failed to complete its action. A correct user program, then, had to check the error code to determine whether to simply try the system routine again. The MIT guy did not like this solution because it was not the right thing. &mdash; <a href="http://ift.tt/1ywca32">The Rise of &laquo;Worse is Better&raquo; By Richard Gabriel</a></p>
</blockquote>
<p>In other words, if you&rsquo;ve captured Ctrl+C from the user&rsquo;s input, then the operating system, instead of just calling your handler, will interrupt the syscall that was running before and return <code>EINTR</code> error code from the kernel. As a result, the programmer will have to foresee this <code>EINTR</code> and that complicates the userspace code. At the cost of simplifying the kernel code. Yes, it should have been done differently &ndash; complicate the kernel code and simplify the userspace code. But the guy from the quote above didn&rsquo;t give a damn about it. He actually said &ldquo;I don&rsquo;t care that everyone will suffer. The main thing is that the kernel code is simpler&rdquo;.</p>
<p>It took off from there. The mentioned behaviour was later fixed in UNIX systems by adding the so-called <code>SA_RESTART</code>. So, they added a special flag instead of fixing everything. Not only have they added <code>SA_RESTART</code>, it does not always work! In particular, in GNU/Linux <code>select</code>, <code>poll</code>, <code>nanosleep</code> and others do not continue their operation after the captured interrupt even in case of <code>SA_RESTART</code>!</p>
<ul>
<li>
<p>In general, specific circumstances that have arisen during the development of the original UNIX, had much influence on it. I&rsquo;ve read somewhere that the <code>cp</code> command is called <code>cp</code> not because of copy but because UNIX was developed with the use of terminals that output characters very slowly. Therefore, it was faster to type <code>cp</code> than <code>copy</code>. I couldn&rsquo;t find the link I saw a long time ago with the example of <code>cp</code> and <code>copy</code>, but here&rsquo;s <a href="http://ift.tt/2lj81DQ">this link</a>. &gt; Commands &mdash; Are These Real Words? &gt; The basic AIX commands (and all UNIX system commands) are, for the most &gt; part, very short, cryptic, two-letter command names. Imagine back years ago, &gt; when computers had only very slow teletype keyboards and paper &ldquo;displays.&rdquo; &gt; (Some of us aren&rsquo;t imagining, we&rsquo;re remembering!) Imagine also, people who &gt; didn&rsquo;t like typing long commands because there was such a long delay between &gt; commands and the computer response. If there were any mistakes, the user had &gt; to retype the whole thing (especially aggravating for folks that type with only &gt; two fingers!). &gt; Also, some UNIX commands came from university students and researchers who &gt; weren&rsquo;t bound by usability standards (no rules, merely peer pressure). They &gt; could write a very useful, clever command and name it anything&mdash;their own initials, &gt; for example (awk by Aho, Weinberger, and Kernighan), or an acronym &gt; (yacc, Yet Another Compiler-Compiler).</p>
</li>
<li>
<p>The names of UNIX utilities is another story. For example, <code>grep</code> comes from command <code>g/re/p</code> in the <code>ed</code> text editor. Well, <code>cat</code> comes from concatenation. I hope you already knew it. To top it all up, <code>vmlinuz</code> &mdash; gZipped LINUx with Virtual Memory support.</p>
</li>
<li>
<p><strong>Out of the blue</strong>, <code>printf</code> is far from being the fastest way of displaying information on the screen or in a file. You didn&rsquo;t know this, did you? The thing is that <code>printf</code>, as well as UNIX itself, was created not for the optimization of time but for the optimization of memory. <code>printf</code> parses a format string in the runtime every single time. That&rsquo;s exactly why a special preprocessor has been invented on the H2O web server that shifts parsing of a format string to the compilation stage. The source can be found <a href="http://ift.tt/1s7kSod">here</a>.</p>
</li>
<li>
<p>When Ken Thompson, the author of UNIX (together with Dennis Ritchie), was asked what he would change in UNIX, he replied that he would name the <code>creat</code> (sic!) function as <code>create</code>. There are lots of sources, e.g. <a href="http://ift.tt/2liWyEp">this one</a>. No comments. I should note that Ken Thompson together with other developers of the original UNIX have created the Plan 9 system that fixes lots of UNIX flaws. This function is called <code>create</code> in it :) He did it :)</p>
</li>
<li>
<p>Here&rsquo;s another quote:</p>
</li>
</ul>
<blockquote>
<p>A child which dies but is never waited for is not really gone in that it still consumes disk swap and system table space. This can make it impossible to create new processes. The bug can be noticed whenseveral &amp; separators are given to the shell not followed by ancommand without an ampersand. Ordinarily things clean themselves upwhen an ordinary command is typed, but it is possible to get into asituation in which no commands are accepted, so no waits are done;the system is then hung.The fix, probably, is to have a new kind of fork which creates aprocess for which no wait is necessary (or possible); also to limit the number of active or inactive descendants allowed to a process. &mdash; <a href="http://ift.tt/2lytZOf">The Source</a></p>
</blockquote>
<p>This quote is from the early UNIX manual. Even then, the existence of zombie processes was considered a bug. But this bug was later simply forgotten. Needless to say, this problem was later solved anyway. So, there are tools in modern GNU/Linux for killing zombie processes, but few people know about them. You can&rsquo;t get rid of a zombie with a regular kill. Everybody says about the existence of zombie processes that &ldquo;It&rsquo;s for design&rdquo;.</p>
<ul>
<li>
<p>Let&rsquo;s talk a bit more about the mentioned C language. Actually, C was developed at the same time with UNIX. That&rsquo;s why when criticizing UNIX, we should comment on C as well. There&rsquo;re lots of things written about how bad C is, and I&rsquo;m not going to repeat all of them. The syntax of types is bad, the processor is terrible, all these <code>4["string"]</code>, and <code>sizeof ('a') != sizeof (char)</code> (in C, not in C++!), and all these <code>i++ + ++i</code>, and <code>while (*p++ = *q++) ;</code> (the example from Stroustrup&rsquo;s book, second revised edition), and so on and so forth.<br><br>
There&rsquo;s just one thing I&rsquo;d like to say. We still don&rsquo;t know how to work conveniently with strings in C. The inconvenience of working with strings always leads to various security issues. This problem is still not solved! Here&rsquo;s a <a href="http://ift.tt/1O3Gv2L">relatively recent document</a> from the committee of C. It discusses quite a questionable way of solving the problem with strings. The conclusion is that this method is bad. Year of publication: 2015. This means that there wasn&rsquo;t a final solution even by 2015!<br><br>
Say nothing of the lack of a simple, user-friendly and multiplatform build system (not this <code>autotools</code> monster that does not support Windows, and not another <code>cmake</code> monster that supports Windows but it&rsquo;s still a monster), the standard package manager, and a user-friendly, like <code>npm</code> (js) or <code>carge</code> (rust), portability library, with the help of which one could at least read content of the folder across all platforms, and at least <em>the main website of C</em> that would be the main entry point for all beginners and would contain not only documentation but also a brief manual on installing C tools on any platform, as well as a manual on creating a simple project in C, and would also contain a user-friendly list of C packages (that should be in the standard repository), and, most importantly, it would be a gathering place for the <em>user community</em>. I&rsquo;ve even registered <em>c-language.org</em> domain hoping to create such a website there. Yeah, dream on! (I even have <em>cpp-language.org</em>, bwahaha!). But they don&rsquo;t have all of this, even though all popular languages have it, except for C and C++. Even Haskell has this! And Rust!<br><br>
In Rust, this jackanapes, that is aimed at the same niche as C, there&rsquo;s a single config that is also a project config, a builder config, and a config for the package manager (actually, <code>cargo</code> is a package manager and a build system at the same time). As a dependency for the given package, it is possible to specify another package located somewhere in <strong>GIT</strong>, including specifying a <strong>GITHUB</strong> repository as a direct dependency. An out-of-the-box support for documentation generation from markdown comments in source code. A package manager that uses <strong>SEMVER</strong> for versions. So, <strong>GIT</strong>, <strong>GITHUB</strong>, <strong>MARKDOWN</strong>, <strong>SEMVER</strong>. In other words, <strong>BUZZWORDS</strong>, <strong>BUZZWORDS</strong> and <strong>HIPSTERS&rsquo; BUZZWORDS</strong>. All of this out-of-the-box. You just go to their main website, and here it is on a silver platter. All of this works the same way on all platforms. Despite the fact that Rust is a system programming language, and not just some JavaScript. Despite the fact that we can play bytes in Rust. There&rsquo;s also a pointer arithmetic in it. So why do Rusters have all of these hipster buzzwords, and we, C-guys, don&rsquo;t? What a shame.<br><br>
I remember how one of my friends asked me where to look for list of packages for C/C++. I had to tell him that there&rsquo;s no such place. He asked me whether C/C++ programmers must suffer. I had nothing to tell him.<br>
Oh, right. I forgot one more thing. Take a look at a prototype of a signal function in the form we see it in the C standard:</p>
<pre>
<code>void (*signal(int sig, void (*func)(int)))(int);
</code>
</pre>
<p>Try to understand it.</p>
</li>
<li>
<p>Terminals in UNIX &mdash; weird legacy. The details are <a href="http://ift.tt/1vc9oNz">here</a>.</p>
</li>
<li>
<p>Filenames in Unix file systems (ext2 and others) are simply a stream of bytes with no encoding. It depends on the locale what encoding they will be interpreted in. So, if we create a file in the operating system in one locale, and then try to take a look at its name in another locale, nothing good will come out of it. There&rsquo;s no such problem in the Windows NTFS.</p>
</li>
<li>
<p>UNIX shell is worse than PHP! Yes, it is, didn&rsquo;t you know? It&rsquo;s a popular thing nowadays to criticize PHP. But UNIX shell is even worse&hellip; It becomes especially bad when we try to develop in it, as it&rsquo;s not a full-fledged programming language. But it&rsquo;s no good even for its niche, scripting common administrative tasks. The reasons for it are shell primitivity, general inefficient arrangement, legacy, tons of special cases, dirty hacks, a complete mess with quotation marks, backslashes, special characters and shell&rsquo;s obsessiveness (just like the entire UNIX ) with plain text.</p>
<ul>
<li>Let&rsquo;s begin with a teaser. How can we recursively find all the files with <code>\</code> name in a folder <em>foo</em>? The correct answer is: <code>find foo -name '\\\\'</code>. We can also do it like this: <code>find foo -name \\\\\\\\</code>. The latter way will cause lots of questions. Try to explain to a person who is not good at UNIX shell why exactly four backslashes are necessary here, not two or eight. We need to write four backslashes here as UNIX shell performs backslash expanding, and <code>find</code> does it too.</li>
<li>How to <code>touch</code> all files in <em>foo</em> (and its subfolders)? At first glance, we could do it like this: <code>find foo | while read A; do touch $A; done</code>. Well, at first glance. Actually, we can come up with 5 things that can ruin it all (and lead to security problems):
<ul>
<li>Filename can contain a backslash. Therefore, we should write <code>read -r A</code> instead of <code>read A</code>
</li>
<li>Filename can contain a slash. That&rsquo;s why we should write <code>touch "$A"</code> instead of <code>touch $A</code>
</li>
<li>Filename can not only contain a space but also start with a space. So we need to write <code>IFS="" read -r A</code> instead of <code>read -r A_</code>
</li>
<li>Filename can contain a newline, so we should use <code>find foo -print0</code> and instead of <code>IFS="" read -r A</code> use <code>IFS="" read -rd ""</code> (I&rsquo;m not really sure here)</li>
<li>Filename can start with a hyphen, so we need to write <code>touch -- "$A"</code> instead of <code>touch "$A"</code>. The final version looks like this: <code>bash find foo -print0 | while IFS="" read -rd "" A; do touch -- "$A"; done</code> Cool, isn&rsquo;t it? By the way, we didn&rsquo;t take into account that POSIX does not guarantee that <code>touch</code> supports option <code>--</code>. Considering this fact, we&rsquo;ll have to check each file on whether it starts with a hyphen (or that it does not start with a slash) and add <code>./</code> to the beginning. Do you understand now why <em>configure</em> scripts generated by <code>autoconf</code> are so large and difficult to read? Because <em>configure</em> needs to take into account all of this crap, including compatibility with various shells. In this example, I used the solution with pipe and loop. I could also use the solution with <code>exec</code> or <code>xargs</code>, but it wouldn&rsquo;t be so eye-catching. (Well, okay. We know that the filename starts with <em>foo</em>, so it cannot start with a space of hyphen).</li>
</ul>
</li>
<li>Let&rsquo;s say we need to delete a file from a variable <code>A</code> on host <em>a@a</em>. How can we do it? Perhaps, like this: <code>ssh a@a rm -- "$A"</code>? (As you might have noticed, we have already taken into account that the filename can contain spaces and start with a hyphen) Never ever do this! <code>ssh</code> is not <code>chroot</code>, or <code>setsid</code>, or <code>nohup</code>, or <code>sudo</code> or any other command that receives an exec-command (meaning a command for direct transmission of the <code>execve</code> family by system calls. <code>ssh</code> (just like <code>su</code>) receives a shell-command, i.e. a command for processing by shell (the exec-command and shell-command are of my own). <code>ssh</code> combines all the arguments into a string, and passes the string to the remote side and performs by shell there. Okay, maybe like this: <code>ssh a@a 'rm &mdash; "$A"'</code>? No, this command tries to find variable <code>A</code> on the remote side. But it&rsquo;s not there, as variables cannot be passed via <code>ssh</code>. Well, maybe like this: <code>ssh a@a "rm -- '$A'"</code>? Nope, this won&rsquo;t work if the filename contains a single quote. Anyway, the correct answer is: <code>ssh a@a "rm -- $(printf '%q\n' "$A")"</code> Convenient, don&rsquo;t you think?</li>
<li>How to get to host <em>a@a</em>, and then to <em>b@b</em> from it, then to <em>c@c</em>, and then to <em>d@d</em> and delete the <em>/foo</em> file from it? Well, this one is simple: <code>bash ssh a@a "ssh b@b \"ssh c@c \\\"ssh d@d \\\\\\\"rm /foo\\\\\\\"\\\"\""</code> Too many backslashes, huh? Well, if you don&rsquo;t like it, let&rsquo;s alternate single and double quotation marks: <code>bash ssh a@a 'ssh b@b "ssh c@c '\''ssh d@d \"rm /foo\"'\''"'</code> By the way, if we were to use Lisp instead of shell, and the <code>ssh</code> function would pass not a string but a parsed AST (abstract syntax tree) to the remote side, there wouldn&rsquo;t be so many backslashes: <code>lisp (ssh "a@a" '(ssh "b@b" '(ssh "c@c" '(ssh "d@d" '(rm "foo")))))</code> &ldquo;Huh? What? Lisp? What Lisp?&rdquo; Curious, aren&rsquo;t you? Go read <a href="http://ift.tt/1jnmOD0">here</a>. You can also refer to other articles by Paul Graham.</li>
<li>Let&rsquo;s combine the previous two paragraphs. A name of the file is in a variable <code>A</code>. We need to go to <em>a@a</em>, and then to <em>b@b</em>, then to<em>c@c</em>, <em>d@d</em> and delete the file in variable <code>A</code>. I&rsquo;m going to leave it for you as an exercise. (I don&rsquo;t know how to do it. :) Well, I might if I thought about it)</li>
<li>
<code>echo</code> is sort of designed for displaying strings on the screen. But the thing is, we can&rsquo;t use it for this purpose if the string is a bit more complex than &ldquo;Hello, world!&rdquo; The only true way to print a random string (e.g. from variable A) is like this: <code>printf '%s\\n' "$A"</code>.</li>
<li>Suppose you want to direct stdout and stderr cmd commands to <em>/dev/null</em>. The riddle: which of these six commands perform the task? <code>bash cmd &gt; /dev/null 2&gt;&amp;1 cmd 2&gt;&amp;1 &gt; /dev/null { cmd &gt; /dev/null; } 2&gt;&amp;1 { cmd 2&gt;&amp;1; } &gt; /dev/null ( cmd &gt; /dev/null ) 2&gt;&amp;1 ( cmd 2&gt;&amp;1 ) &gt; /dev/null</code> Turns out, the correct answer is: the 1st, the 4th and the 6th. And the 2nd, the 3rd, and the 5th don&rsquo;t. And again, I&rsquo;m leaving it to you to figure out the reason as an exercise. :)</li>
</ul>
</li>
<li>
<p>Actually, this post appeared as a response to <a href="http://ift.tt/2kSPuNh">this one</a>. It says that a special date is used in Windows as a driver timestamp. Instead of introducing a special attribute or checking the manufacturer. There&rsquo;re lots of similar things in UNIX. The file is hidden only based on a dot at the beginning of the file instead of the special attribute. I was shocked when I first learnt about it (yeah, in those old times I installed Ubuntu for the first time). &ldquo;What an idiots!&rdquo;, I thought. But I&rsquo;m used to it now. But thinking about it, it&rsquo;s a terrible workaround. Then, shell decides whether it is a login shell based on the hyphen passed by the first character to argv<a href="http://ift.tt/2lZ8F8c?!">0</a>. This is a misuse of <code>argv[0]</code>. <code>argv[0]</code> is not meant for this purpose. Any other method would be better, e.g. using another argument or some environment variable.</p>
</li>
<li>
<p>In BSD sockets, user must change the byte order of the port number on its own. And all because someone has made a mistake in the UNIX kernel code, missing to foresee the byte order change. As a temporary hack, this someone fixed the user space instead of the kernel code. That&rsquo;s how it goes. That&rsquo;s where it came from to Windows (together with file <code>/etc/hosts</code>, aka <code>C:\windows\system32\drivers\etc\hosts</code>). <a href="http://ift.tt/2lyqbwF">The Source</a>.</p>
</li>
</ul>
<h2 id="unix-philosophy">UNIX Philosophy</h2>
<p>Some people think that UNIX is great and perfect, and that all its basic ideas (&laquo;everything is a file&raquo;, &laquo;everything is text&raquo; and so on) are amazing and form the so-called &rdquo;UNIX Philosophy&rdquo;. I guess you&rsquo;re starting to understand that it&rsquo;s not quite so. Let&rsquo;s review this &ldquo;Unix philosophy&rdquo;. Have a look at some points below. I&rsquo;m not trying to say that all of these things should be canceled, I&rsquo;m simply pointing at some drawbacks. * &ldquo;Everything is text&rdquo;. As we&rsquo;ve already seen in the example with <code>/etc/passwd</code>, the widespread use of plain text can lead to performance problems. UNIX authors have actually invented a format for each system config (<code>passwd</code>, <code>fstab</code>, etc.). With their rules of escaping special characters. Surprised? <code>/etc/fstab</code> uses spaces and line breaks as separators. But what if folder names include, say, spaces? For this case, the format of <code>fstab</code> provides special escape characters for folder names. Turns out, any script reading <code>fstab</code> should be able to interpret the escape character. For example, the <code>fstab-decode</code> utility meant for this purpose (run as root). You didn&rsquo;t know this, did you? Go fix your scripts. :) As a result, we need a parser for each system config. It would be much easier if we used JSON or XML for system configs. Or maybe some binary format. Especially for those configs that are constantly read by different programs. As a result, they need a good read rate (it&rsquo;s higher in binary formats).<br><br>
That&rsquo;s not all I wanted to say about &ldquo;everything is text&rdquo;. Standard utilities provide the output in the form of a plain text. For each utility, we actually need a parser of its own. We often need to parse an output by using <code>sed</code>, <code>grep</code>, <code>awk</code>, etc. Each utility has its own options to determine what columns to display, what columns to sort by, and so on. It would be better if utilities provided the output in the form of XML, JSON, some binary format or anything else. To display this information in a user-friendly way and to work further with it, we could pipe the result to additional utilities that remove some columns, sort by some column, select the required strings, etc. They either display the result in the form of a nice table or pass it somewhere else. All of this is carried out in a multipurpose way that does not depend on the initial utility that generated the output. With no need to parse anything by regex. UNIX shell isn&rsquo;t good at working with JSON and XML. But UNIX shell has plenty of other drawbacks. We&rsquo;d better throw it away and use some other language that can work well with things like JSON and do lots of everything else.<br><br>
Just imagine! Let&rsquo;s say we need to delete all files in the current folder of size bigger than 1 kilobyte. Yes, I know that we can do this with <code>find</code>. But let&rsquo;s suppose we definitely need to do this via <code>ls</code> (and without <code>xargs</code>). How to do it? Like this:</p>
<pre>
<code class="language-bash">LC_ALL=C ls -l | while read -r MODE LINKS USER GROUP SIZE M D Y FILE; do if [ "$SIZE" -gt 1024 ]; then rm -- "$FILE"; fi; done
</code>
</pre>
<p>We need LC_ALL here to be sure that the date will take exactly three words in the output of <code>ls</code>. This solution not only looks ugly, but also has a number of drawbacks. Firstly, it will not work if the file name contains a line break, or begins with a space. Next, we need to explicitly list the names of all <code>ls</code> columns or at least remember where the ones we need (i.e. SIZE and FILE) are located. If we make a mistake in the order of columns, the error will become apparent only during the runtime. When we delete the wrong files. :)<br><br>
How would the solution look like in the perfect world I&rsquo;m suggesting? Something like this: <code>ls | grep 'size &gt; 1kb' | rm</code>. It&rsquo;s short, and, most importantly, you can see the meaning in code and it&rsquo;s impossible to make a mistake. Let&rsquo;s see. In my world, <code>ls</code> always gives all the information. We don&rsquo;t need a special <code>-l</code> option for this. When it&rsquo;s necessary to delete all columns and leave the filename only , we can do this with a special utility we should direct the <code>ls</code> output to. Thus, <code>ls</code> provides a lift of files in some structured form, say, JSON. This representation &ldquo;knows&rdquo; names of columns and their types. i.e. that is a string, a number or something else. Then, this output is piped to <code>grep</code> that, in my world, selects the necessary strings from JSON. JSON &ldquo;knows&rdquo; field names, so <code>grep</code> &ldquo;understands&rdquo; what &ldquo;size&rdquo; means here. Moreover, JSON contains information about the type of size field. It contains information that it&rsquo;s a number, and even that it&rsquo;s not just a number but a file size. Therefore, we can compare it to 1kb. Next, <code>grep</code> pipes the output to <code>rm</code>. <code>rm</code> &ldquo;sees&rdquo; that it&rsquo;s going to receive files. Yes, JSON also stores information about the type of these strings, that they&rsquo;re files. <code>rm</code> deletes them. JSON is also responsible for correct special characters escaping. That&rsquo;s why files with special characters &ldquo;simply work&rdquo;. Cool, right? I took the idea from <a href="http://ift.tt/2liXAjF">here</a>. It should also be mentioned that something of the kind is implemented in Windows Powershell.</p>
<ul>
<li>UNIX shell. Another basic idea of UNIX. I&rsquo;ve already mentioned small disadvantages of UNIX shell in the first part of the article. What&rsquo;s &ldquo;cool&rdquo; about UNIX shell? <em>At the moment of its release</em> (it was a long time ago), it was much stronger than command interpreters embedded in other operating systems. It allowed to write more powerful scripts. Seems like the UNIX shell was the most powerful scripting language when it was released. Because there were no <em>sane</em> scripting languages back in the days (meaning the ones that would allow full-fledged programming and not just scripting). It was later when a programmer named Larry Wall noticed that UNIX shell lacked a lot to be considered a good programming language. He decided to combine the simplicity of UNIX shell with the ability of full-fledged programming from C. He created Perl. Yes, Perl and other following script programming languages actually replaced UNIX shell. Even Rob Pike, one of the authors (to my mind) of &ldquo;UNIX philosophy&rdquo;, confirms the fact. Answering a question about &ldquo;one tool for one job&rdquo; <a href="http://ift.tt/2l6xq24">here</a>, he said: &ldquo;Those days are dead and gone and the eulogy was delivered by Perl&rdquo;. Actually, I believe that this phrase refers to a typical use of UNIX shell, i.e. to a situation of combining a big number of small tools in a shell script. No, says Pike, simply use Perl.<br><br>
I&rsquo;m not done talking about UNIX shell. Let&rsquo;s review the example of shell code I&rsquo;ve already provided above: <code>shell find foo -print0 | while IFS="" read -rd "" A; do touch -- "$A"; done</code> We call <code>touch</code> in the loop here (I do know that the code can be re-written with <code>xargs</code>, so that <code>touch</code> is called only once; but let&rsquo;s forget about it for now, okay?). We call <code>touch</code> in the loop! This means there is a new process for each file. This is extremely inefficient. Code in any other programming language will work faster than this one. But when UNIX shell appeared on the scene, it was one of the few languages that allowed writing this action in one string.<br><br>
Long story short, we should use any other script programming language instead of UNIX shell. A language that will be suitable not only for scripting, but for real programming as well. A language that does not run a new process every time we need to <code>touch</code> a file. Perhaps, we&rsquo;ll have to borrow some features from shell to simplify things even more.</li>
<li>Simplicity. I&rsquo;m not talking about shell and combining lots of simple utilities from shell (it was the previous point). I&rsquo;m talking about simplicity I general. Using simple tools. Say, editing a picture with <code>sed</code>. Yes, yes. To convert <em>jpg</em> into <em>ppm</em> with the help of the command line. Then, edit the picture using the graphic editor, <code>grep</code>, and <code>sed</code>. And then convert it back into <em>jpg</em>. Yes, we can do this. However, it&rsquo;s often better to do this with Photoshop or GIMP, although they&rsquo;re large, integrated programs. Not in the UNIX style.</li>
</ul>
<p>I guess I&rsquo;ll finish adding points now. Yep, that&rsquo;s enough. There&rsquo;re some ideas in UNIX I really like. Like this one: &ldquo;a program should do one thing, and do it well&rdquo;. But not within the context of shell. I guess you&rsquo;ve realized by now that I don&rsquo;t like shell. (Once again, I think that in the interview provided above Rob Pike took the principle &ldquo;a program should do one thing, and do it well&rdquo; in the context of shell and therefore rejected it) I&rsquo;m talking about this principle in its essence. For example, a console mail client should not have a built-in text editor; it should just run some external editor. Or the principle, by which one must write a program&rsquo;s console kernel before its graphical user interface.</p>
<p>Now, the general picture. Once there was UNIX. It was a breakthrough at the time. It was better than its competitors in lots of things. UNIX had a lot of ideas. As any other operating system, UNIX required from programmers compliance with certain principles when writing application programs. The fundamental ideas got the name of &ldquo;UNIX philosophy&rdquo;. One of the people who formulated the UNIX philosophy was the already mentioned Rob Pike. He did this in his presentation titled <a href="http://ift.tt/RnQa8F">&ldquo;UNIX Style, or cat -v Considered Harmful&rdquo;</a>. After the presentation, Rob Pike and Brian Kerninghan published an <a href="http://ift.tt/1s05PI3">article based on the presentation</a>. They told us that, say, the purpose of <code>cat</code> is concatenation and nothing else. Perhaps, Rob Pike was the one to invent the &ldquo;UNIX philosophy&rdquo;. <a href="http://cat-v.org/">cat-v.org</a> website was named after this presentation. Read it, it&rsquo;s a very interesting website.</p>
<p>But then, many years later, Pike made two more presentations, in which, I think, he abandoned his philosophy. You got it, fans? Your idol gave up his own philosophy. You can go home now. In the first presentation<a href="http://ift.tt/1AUwCQB">&ldquo;Systems Software Research is Irrelevant&rdquo;</a> Pike complains that no one writes new operating systems anymore. Even if they do, it&rsquo;s another UNIX: &ldquo;New operating systems today tend to be just ways of reimplementing Unix. If they have a novel architecture &mdash; and some do &mdash; the first thing to build is the Unix emulation layer. How can operating systems research be relevant when the resulting operating systems are all indistinguishable&rdquo;</p>
<p>The second Pipe&rsquo;s presentation is titled <a href="http://ift.tt/2l6ux1q">&ldquo;The Good, the Bad, and the Ugly: The Unix Legacy&rdquo;</a>. Pike says that flat text is not multipurpose; it&rsquo;s good but it doesn&rsquo;t always work: &ldquo;What makes the system good at what it&rsquo;s good at is also what makes it bad at what it&rsquo;s bad at. Its strengths are also its weaknesses. A simple example: flat text files. Amazing expressive power, huge convenience, but serious problems in pushing past a prototype level of performance or packaging. Compare the famous spell pipeline with an interactive spell-checker&rdquo;. Then: &ldquo;C hasn&rsquo;t changed much since the 1970s&hellip; And &mdash; let&rsquo;s face it &mdash; it&rsquo;s ugly&rdquo;. Then Pipe admits the limitation of pipes connecting simple utilities, as well as the limitation of regex.</p>
<p>UNIX was genius at the time of its introduction. Especially if we remember what tools the authors of UNIX had. They didn&rsquo;t have a ready-made UNIX to develop UNIX on. They didn&rsquo;t have IDE. They even developed in assembler at the beginning. I guess the only things they had were an assembler and a text editor.</p>
<p>At a certain point, people standing at the origins of UNIX began to write a new operating system: Plan 9. Including Ken Thompson, Dennis Ritchie and Rob Pike. Taking into account the numerous mistakes of UNIX. However, no one raises Plan 9 on a pedestal. Pike mentions Plan 9 in &ldquo;Systems Software Research is Irrelevant&rdquo;, but still encourages us to write new operating systems.</p>
<p>James Hague, a veteran of programming (has been in programming since the eighties), writes the following: &ldquo;What I was trying to get across is that if you romanticize Unix, if you view it as a thing of perfection, then you lose your ability to imagine better alternatives and become blind to potentially dramatic shifts in thinking&rdquo;. <a href="http://ift.tt/witFTc">The Source</a>. Read this article and the one he refers to: <a href="http://ift.tt/1IsArh9">&laquo;Free Your Technical Aesthetic from the 1970s&raquo;</a>. Actually, if you like my article, you&rsquo;ll like his blog too.</p>
<p>So, I do not want to say that UNIX &ndash; is a bad system. I&rsquo;m just drawing your attention to the fact that it has tons of drawbacks, just like other systems do. I also do not cancel the &ldquo;UNIX philosophy&rdquo;, just trying to say that it&rsquo;s not an absolute. My post is mostly referred to UNIX and GNU/Linux fans. The provocative tone has been used just to attract your attention.</p>
</div>
</div>
<br><br><br><br>
Technology<br><br>
via Hacker News http://ift.tt/YV9WJO<br><br>
March 2, 2017 at 07:09AM
